src/Magento/FunctionalTestingFramework/StaticCheck/TestDependencyCheck.php

```diff
@@ -0,0 +1,385 @@
+                }
+            }
+
+            $currentModule = $this->moduleNameToComposerName[$moduleFullName];
+            $modulesReferencedInTest = $this->getModuleDependenciesFromReferences($allEntities, $currentModule);
+            $moduleDependencies = $this->flattenedDependencies[$moduleFullName];
+            // Find Violations
+            $violatingReferences = [];
+            foreach ($modulesReferencedInTest as $entityName => $files) {
+                $valid = false;
+                foreach ($files as $module) {
+                    if (array_key_exists($module, $moduleDependencies) || $module == $currentModule) {
+                        $valid = true;
+                        break;
+                    }
+                }
+                if (!$valid) {
+                    $violatingReferences[$entityName] = $files;
+                }
+            }
+
+            if (!empty($violatingReferences)) {
+                // Build error output
+                $errorOutput = "\nFile \"{$filePath->getRealPath()}\"\ncontains references to following modules:\n\t\t";
+                foreach ($violatingReferences as $entityName => $files) {
+                    $errorOutput .= "\n\t {$entityName} from modules: " . implode(", ", $files);
+                }
+                $testErrors[$filePath->getRealPath()][] = $errorOutput;
+            }
+        }
+        return $testErrors;
+    }
+
+    /**
+     * Builds and returns array of FullModuleNae => composer name
+     * @param array $array
+     * @return array
+     */
+    private function buildModuleNameToComposerName($array)
+    {
+        $returnList = [];
+        foreach ($array as $moduleName => $path) {
+            $composerData = json_decode(file_get_contents($path . DIRECTORY_SEPARATOR . "composer.json"));
+            $returnList[$moduleName] = $composerData->name;
+        }
+        return $returnList;
+    }
+
+    /**
+     * Builds and returns flattened dependency list based on composer dependencies
+     * @return array
+     */
+    private function buildComposerDependencyList()
+    {
+        $flattenedDependencies = [];
+
+        foreach ($this->moduleNameToPath as $moduleName => $pathToModule) {
+            $composerData = json_decode(file_get_contents($pathToModule . DIRECTORY_SEPARATOR . "composer.json"), true);
+            $this->allDependencies[$moduleName] = $composerData['require'];
+        }
+        foreach ($this->allDependencies as $moduleName => $dependencies) {
+            $this->alreadyExtractedDependencies = [];
+            $flattenedDependencies[$moduleName] = $this->extractSubDependencies($moduleName);
+        }
+        return $flattenedDependencies;
+    }
+
+    /**
+     * Recursive function to fetch dependencies of given dependency, and its child dependencies
+     * @param string $subDependencyName
+     * @return array
+     */
+    private function extractSubDependencies($subDependencyName)
```

Comment:
This recursive function is fine by itself. But any reason that you don't build the recursion from root level composer file. Will the current way cause duplicate dependencies in the root level?

Reply:
I **think** the reason is because it would flatten all the dependencies to one huge array, and we want an array with several `module => [dependencies]`.
We iterate through the big list of dependencies on ln 283 so that we can scope the above flattening.